bank-projects.cue

```diff
@@ -0,0 +1,103 @@
+@if(!NoBank)
+package holos
+
+// Platform wide configuration.
+BankOfHolos: #BankOfHolos & {
+	Name: string | *"bank-of-holos"
+
+	Environments: {
+		dev: tier:   "nonprod"
+		test: tier:  "nonprod"
+		stage: tier: "nonprod"
+		prod: tier:  "prod"
+	}
+
+	EnvironmentNamespaces: {
+		"bank-frontend": _
+		"bank-backend":  _
+		"bank-security": _
+	}
+}
+
+// Projects are security boundaries, so manage one project for each environment
+// and team combination.
+for ENV in BankOfHolos.configuration.environments {
```

Comment:
This where we tie together the concepts of environments, clusters, and a stack of components into a project.  The project is intended as a security boundary, so our example bank stack gets broken down into 3 projects for each deployment environment.  The backend dev team are responsible for their dev,test,stage,prod projects.

The web team and security team have respective projects as well.

It can help to visualize the concrete values this results in:

```
holos cue export . --expression BankOfHolos.configuration --out=yaml
```

```yaml
environments:
  dev:
    name: dev
    namespaces:
      dev-bank-frontend:
        metadata:
          name: dev-bank-frontend
          labels:
            kubernetes.io/metadata.name: dev-bank-frontend
        kind: Namespace
        apiVersion: v1
      dev-bank-backend:
        metadata:
          name: dev-bank-backend
          labels:
            kubernetes.io/metadata.name: dev-bank-backend
        kind: Namespace
        apiVersion: v1
      dev-bank-security:
        metadata:
          name: dev-bank-security
          labels:
            kubernetes.io/metadata.name: dev-bank-security
        kind: Namespace
        apiVersion: v1
    frontend:
      namespace: dev-bank-frontend
    backend:
      namespace: dev-bank-backend
    security:
      namespace: dev-bank-security
  test:
    name: test
    namespaces:
      test-bank-frontend:
        metadata:
          name: test-bank-frontend
          labels:
            kubernetes.io/metadata.name: test-bank-frontend
        kind: Namespace
        apiVersion: v1
      test-bank-backend:
        metadata:
          name: test-bank-backend
          labels:
            kubernetes.io/metadata.name: test-bank-backend
        kind: Namespace
        apiVersion: v1
      test-bank-security:
        metadata:
          name: test-bank-security
          labels:
            kubernetes.io/metadata.name: test-bank-security
        kind: Namespace
        apiVersion: v1
    frontend:
      namespace: test-bank-frontend
    backend:
      namespace: test-bank-backend
    security:
      namespace: test-bank-security
  stage:
    name: stage
    namespaces:
      stage-bank-frontend:
        metadata:
          name: stage-bank-frontend
          labels:
            kubernetes.io/metadata.name: stage-bank-frontend
        kind: Namespace
        apiVersion: v1
      stage-bank-backend:
        metadata:
          name: stage-bank-backend
          labels:
            kubernetes.io/metadata.name: stage-bank-backend
        kind: Namespace
        apiVersion: v1
      stage-bank-security:
        metadata:
          name: stage-bank-security
          labels:
            kubernetes.io/metadata.name: stage-bank-security
        kind: Namespace
        apiVersion: v1
    frontend:
      namespace: stage-bank-frontend
    backend:
      namespace: stage-bank-backend
    security:
      namespace: stage-bank-security
  prod:
    name: prod
    namespaces:
      prod-bank-frontend:
        metadata:
          name: prod-bank-frontend
          labels:
            kubernetes.io/metadata.name: prod-bank-frontend
        kind: Namespace
        apiVersion: v1
      prod-bank-backend:
        metadata:
          name: prod-bank-backend
          labels:
            kubernetes.io/metadata.name: prod-bank-backend
        kind: Namespace
        apiVersion: v1
      prod-bank-security:
        metadata:
          name: prod-bank-security
          labels:
            kubernetes.io/metadata.name: prod-bank-security
        kind: Namespace
        apiVersion: v1
    frontend:
      namespace: prod-bank-frontend
    backend:
      namespace: prod-bank-backend
    security:
      namespace: prod-bank-security
```